ledger-core/testutils/cloud/controller_test.go

```diff
@@ -0,0 +1,161 @@
+// Copyright 2020 Insolar Network Ltd.
+// All rights reserved.
+// This material is licensed under the Insolar License version 1.0,
+// available at https://github.com/insolar/assured-ledger/blob/master/LICENSE.md.
+
+// +build slowtest
+
+package cloud_test
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/require"
+
+	"github.com/insolar/assured-ledger/ledger-core/application/api/requester"
+	"github.com/insolar/assured-ledger/ledger-core/application/testutils/launchnet"
+	"github.com/insolar/assured-ledger/ledger-core/configuration"
+	"github.com/insolar/assured-ledger/ledger-core/instrumentation/insapp"
+	"github.com/insolar/assured-ledger/ledger-core/pulse"
+	"github.com/insolar/assured-ledger/ledger-core/reference"
+	"github.com/insolar/assured-ledger/ledger-core/server"
+	"github.com/insolar/assured-ledger/ledger-core/testutils"
+	"github.com/insolar/assured-ledger/ledger-core/testutils/cloud"
+)
+
+func TestController_PartialDistribute(t *testing.T) {
+	var (
+		numVirtual        = 10
+		numLightMaterials = 0
+		numHeavyMaterials = 0
+	)
+
+	cloudSettings := launchnet.CloudSettings{Virtual: numVirtual, Light: numLightMaterials, Heavy: numHeavyMaterials}
+
+	appConfigs, cloudBaseConf, certFactory, keyFactory := launchnet.PrepareCloudConfiguration(cloudSettings)
+	baseConfig := configuration.Configuration{}
+	baseConfig.Log = cloudBaseConf.Log
+
+	confProvider := &server.CloudConfigurationProvider{
+		PulsarConfig:       cloudBaseConf.PulsarConfiguration,
+		BaseConfig:         baseConfig,
+		CertificateFactory: certFactory,
+		KeyFactory:         keyFactory,
+		GetAppConfigs: func() []configuration.Configuration {
+			return appConfigs
+		},
+	}
+
+	controller := cloud.NewController()
+	s := server.NewControlledMultiServer(controller, confProvider)
+	go func() {
+		s.Serve()
+	}()
+
+	// wait for starting all components
+	for !s.(*insapp.Server).Started() {
```

Comment:
this looks ugly. We need to either broaden Server interface or return *insapp.Server so we don't need this